File: tagalong/018-inline-functions/inline-functions.go
package main

import "fmt"

func main() {
	a := 1
	notSoRand := func() int {
		a = a * 7
		return a
	}
	call1 := notSoRand()
	call2 := notSoRand()
	fmt.Println("calling function yields different results:", call1, call2)

	superrand1 := SuperRandom(notSoRand)
	superrand2 := SuperRandom(staticRandom)
	fmt.Println("a function can take another function as argument:", superrand1, superrand2)
}

// SuperRandom returns a super random number using a
// not so random source function "normalRandom".
// (this is not actually more random!)
func SuperRandom(normalRandom func() int) int {
	superrand := 12345678
	for i := 0; i < 3; i++ {
		rand := normalRandom()
		superrand = superrand*7 + rand*31
	}
	return superrand
}

func staticRandom() int {
	return 4
}

File: tagalong/018-inline-functions/inline-functions.py
def SuperRandom(normalRandom:lambda:int) -> int:
    """SuperRandom returns a super random number using a
    not so random source function "normalRandom".
    (this is not actually more random!)"""
    superrand = 12345678
    for _ in range(3):
        rand = normalRandom()
        superrand = superrand*7 + rand*31
    return superrand

# Python's inline function creation uses
# the lambda keyword, though its limited to
# one-liners so we can't really use it here.
# We do with a simple def.
a = 287117
def notSoRand()->int:
    global a
    a = a * 7
    return a

superrand1 = SuperRandom(notSoRand)
superrand2 = SuperRandom(notSoRand)
superrand3 = SuperRandom(notSoRand)

print(superrand1, superrand2, superrand3)




